lib/frame.js

```diff
@@ -90,44 +104,79 @@ api.frame = (state, subjects, frame, parent, property = null) => {
     result, clear the unique embedded subjects map when the property is null,
     which only occurs at the top-level. */
     if(property === null) {
-      state.uniqueEmbeds = {};
+      state.uniqueEmbeds = {[state.graph]: {}};
+    } else {
+      state.uniqueEmbeds[state.graph] = state.uniqueEmbeds[state.graph] || {}
     }
 
     // start output for subject
     const output = {};
     output['@id'] = id;
+    if(id.indexOf('_:') === 0) {
+      util.addValue(state.bnodeMap, id, output, {propertyIsArray: true});
+    }
     state.link[id] = output;
 
     // if embed is @never or if a circular reference would be created by an
     // embed, the subject cannot be embedded, just add the reference;
     // note that a circular reference won't occur when the embed flag is
     // `@link` as the above check will short-circuit before reaching this point
     if(flags.embed === '@never' ||
-      _createsCircularReference(subject, state.subjectStack)) {
+      _createsCircularReference(subject, state.graph, state.subjectStack)) {
       _addFrameOutput(parent, property, output);
       continue;
     }
 
     // if only the last match should be embedded
     if(flags.embed === '@last') {
       // remove any existing embed
-      if(id in state.uniqueEmbeds) {
+      if(id in state.uniqueEmbeds[state.graph]) {
         _removeEmbed(state, id);
       }
-      state.uniqueEmbeds[id] = {parent: parent, property: property};
+      state.uniqueEmbeds[state.graph][id] = {parent: parent, property: property};
     }
 
     // push matching subject onto stack to enable circular embed checks
-    state.subjectStack.push(subject);
+    state.subjectStack.push({subject: subject, graph: state.graph});
+
+    // subject is also the name of a graph
+    if(id in state.graphMap) {
+      let recurse = false;
+      let subframe = null;
+      if(!('@graph' in frame)) {
+        recurse = state.graph !== '@merged';
+        subframe = {};
+      } else {
+        subframe = frame['@graph'][0];
+        if(!types.isObject(subframe)) {
+          subframe == {};
```

Comment:
This is clearly not correct.  Should a test cover this somehow?

Reply:
It's CYA code, as it should never be the case that `@graph` doesn't have an array of object values. But, perhaps it simply can't happen because of sanity checks in expansion.